cmd/ncproxy/service.go

```diff
@@ -0,0 +1,211 @@
+package main
+
+import (
+	"log"
+	"os"
+	"time"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/debug"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+const serviceName = "ncproxy"
+
+var (
+	panicFile *os.File
+	oldStderr windows.Handle
+)
+
+type handler struct {
+	fromsvc chan error
+	done    chan struct{}
+}
+
+type serviceFailureActions struct {
+	ResetPeriod  uint32
+	RebootMsg    *uint16
+	Command      *uint16
+	ActionsCount uint32
+	Actions      uintptr
+}
+
+type scAction struct {
+	Type  uint32
+	Delay uint32
+}
+
+// See http://stackoverflow.com/questions/35151052/how-do-i-configure-failure-actions-of-a-windows-service-written-in-go
+const (
+	scActionNone    = 0
+	scActionRestart = 1
+
+	serviceConfigFailureActions = 2
+)
+
+func initPanicFile(path string) error {
+	panicFile, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
+	if err != nil {
+		return err
+	}
+
+	st, err := panicFile.Stat()
+	if err != nil {
+		return err
+	}
+
+	// If there are contents in the file already, move the file out of the way
+	// and replace it.
+	if st.Size() > 0 {
+		panicFile.Close()
+		_ = os.Rename(path, path+".old")
+		panicFile, err = os.Create(path)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Update STD_ERROR_HANDLE to point to the panic file so that Go writes to
+	// it when it panics. Remember the old stderr to restore it before removing
+	// the panic file.
+	sh := uint32(windows.STD_ERROR_HANDLE)
+	h, err := windows.GetStdHandle(sh)
+	if err != nil {
+		return err
+	}
+	oldStderr = h
+
+	if err := windows.SetStdHandle(sh, windows.Handle(panicFile.Fd())); err != nil {
+		return err
+	}
+
+	// Reset os.Stderr to the panic file (so fmt.Fprintf(os.Stderr,...) actually gets redirected)
+	os.Stderr = os.NewFile(panicFile.Fd(), "/dev/stderr")
+
+	// Force threads that panic to write to stderr (the panicFile handle now).
+	log.SetOutput(os.Stderr)
+	return nil
+}
+
+func removePanicFile() {
+	if st, err := panicFile.Stat(); err == nil {
+		if st.Size() == 0 {
```

Comment:
what does it mean for the size here to not be 0?

Reply:
There's data in the panic file, e.g. we had a panic.

Reply:
maybe add a comment about this?

Reply:
Sure!